### PR TITLE
zwe components package registry handling

### PR DESCRIPTION
**Pull request by 1000TurquoisePogs**
This is an enhancement to zwe that brings new "zwe components" commands. It does not change existing behavior and is opt-in via zowe.useConfigmgr=true and defining a zowe.extensionRegistry object in zowe.yaml
For more info, please see the attached presentation
[appstore2.pdf](https://github.com/zowe/zowe-install-packaging/files/9292283/appstore2.pdf)

demo recording: https://zoom.us/rec/share/y6zsW5U9QWE1s1r4M3nFnSO9Kkv3yeT5boyZFqWH1BxW3Tju_jcAGP7jO1DsLuZq.rhlqHx6DgPxmXBhW?startTime=1660053548000

### CODE REVIEW

**Comment by 1000TurquoisePogs** [comment]
Did a demo and saw bugs in "update" and conda, which were working before, so I'll make some fixes today.

**Comment by armstro** [comment]
Sean - I just listened to the recording of "App Store" Innovation item (or maybe best called npm/conda delivery and install of Zowe extensions on z/OS with zwe?..but App Store rolls off the tongue easier ...). 

- I did hear you say the repo could reside on customer premise for those customers that don't want to download from public repo 
- I did not hear enough about how "trusted" packages of software can be checked to be unaltered. Do npm/conda already have a mechanism to check with some authoritative source for package authenticity?

**Comment by 1000TurquoisePogs** [comment]
> Sean - I just listened to the recording of "App Store" Innovation item (or maybe best called npm/conda delivery and install of Zowe extensions on z/OS with zwe?..but App Store rolls off the tongue easier ...).
> 
>     * I did hear you say the repo could reside on customer premise for those customers that don't want to download from public repo
> 
>     * I did not hear enough about how "trusted" packages of software can be checked to be unaltered. Do npm/conda already have a mechanism to check with some authoritative source for package authenticity?

npm and conda both download via HTTPS, and checksum their packages so that they can know the package is unaltered from their point of view, but because that checksum is also visible to us, we can check it versus what a vendor said their npm/conda package hash should be. The pax file of an extension also doesnt need to be modified in this design, so when it comes out the other side of npm or conda, the pax also can be hashed against what's expected.

In a UI... the missing part of an "app store", we could display those hashes for easy confirmation before the user continues. I could add that to the CLI as well.

File: bin/commands/components/install/.exclusive-parameters [hunk 1]
```diff
@@ -1,3 +1,5 @@
-component-file,component|o|string|required||||Path to the component package or directory.
+component-file,component|o|string|required||||Either a path or component name. The path must be to a component package or directory. If a name is specified instead, install checks the zowe package registry.
 auto-encoding|e|string||auto|||If we want to automatically tagging the module files.
-skip-enable||boolean|||||Install component without enabling it for use.
\ No newline at end of file
+skip-enable||boolean|||||Install component without enabling it for use.
+registry|r|string|||||Specifies the registry to searh within instead of the default. The registry must be compatible with the manager used.
+manager|m|string|||||Specifies the manager to search the registry with instead of the default. The manager must be compatible with the registry used.
```
Comment (1000TurquoisePogs) [inline bin/commands/components/install/.exclusive-parameters]: change this to handler

File: bin/commands/components/install/index.ts [hunk 2]
```diff
@@ -14,17 +14,78 @@ import * as extract from './extract/index';
...
+      //all extensions doesnt mean every one exists from this handler. handler must check them.
+      component = allExtensions;
+    }
   }
-  if (!skipEnable) {
-    componentEnable.execute(componentName);
+
+  handler = getHandler(handler);
+  if (!handler) {
+    common.printErrorAndExit("Error ZWEL????E: Handler (-handler or zowe.extensionRegistry.defaultHandler) required but not specified", undefined, 255);
```
Comment (1000TurquoisePogs) [inline bin/commands/components/install/index.ts]: assign error code, not just ????

File: bin/commands/components/search/.errors [hunk 3]
```diff
@@ -0,0 +1,3 @@
+ZWEL????E|???|Command requires zowe.useConfigmgr=true to use.
```
Comment (1000TurquoisePogs) [inline bin/commands/components/search/.errors]: fill this in

File: bin/commands/components/search/index.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+USE_CONFIGMGR=$(check_configmgr_enabled)
+if [ "${USE_CONFIGMGR}" = "true" ]; then
+  if [ -z "${ZWE_PRIVATE_TMP_MERGED_YAML_DIR}" ]; then
+    # user-facing command, use tmpdir to not mess up workspace permissions
+    export ZWE_PRIVATE_TMP_MERGED_YAML_DIR=1
+  fi
+  _CEE_RUNOPTS="XPLINK(ON),HEAPPOOLS(OFF)" ${ZWE_zowe_runtimeDirectory}/bin/utils/configmgr -script "${ZWE_zowe_runtimeDirectory}/bin/commands/components/search/cli.js"
+else
+  print_error_and_exit "Error ZWEL????E: Command requires zowe.useConfigmgr=true to use." "" 255
```
Comment (1000TurquoisePogs) [inline bin/commands/components/search/index.sh]: add number

File: bin/commands/components/uninstall/index.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+USE_CONFIGMGR=$(check_configmgr_enabled)
+if [ "${USE_CONFIGMGR}" = "true" ]; then
+  if [ -z "${ZWE_PRIVATE_TMP_MERGED_YAML_DIR}" ]; then
+    # user-facing command, use tmpdir to not mess up workspace permissions
+    export ZWE_PRIVATE_TMP_MERGED_YAML_DIR=1
+  fi
+  _CEE_RUNOPTS="XPLINK(ON),HEAPPOOLS(OFF)" ${ZWE_zowe_runtimeDirectory}/bin/utils/configmgr -script "${ZWE_zowe_runtimeDirectory}/bin/commands/components/uninstall/cli.js"
+else
+  print_error_and_exit "Error ZWEL????E: Command requires zowe.useConfigmgr=true to use." "" 255
```
Comment (1000TurquoisePogs) [inline bin/commands/components/uninstall/index.sh]: add number

File: bin/commands/components/upgrade/index.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+USE_CONFIGMGR=$(check_configmgr_enabled)
+if [ "${USE_CONFIGMGR}" = "true" ]; then
+  if [ -z "${ZWE_PRIVATE_TMP_MERGED_YAML_DIR}" ]; then
+    # user-facing command, use tmpdir to not mess up workspace permissions
+    export ZWE_PRIVATE_TMP_MERGED_YAML_DIR=1
+  fi
+  _CEE_RUNOPTS="XPLINK(ON),HEAPPOOLS(OFF)" ${ZWE_zowe_runtimeDirectory}/bin/utils/configmgr -script "${ZWE_zowe_runtimeDirectory}/bin/commands/components/upgrade/cli.js"
+else
+  print_error_and_exit "Error ZWEL????E: Command requires zowe.useConfigmgr=true to use." "" 255
```
Comment (1000TurquoisePogs) [inline bin/commands/components/upgrade/index.sh]: message id needed

File: manifest.json.template [hunk 7]
```diff
@@ -63,7 +63,7 @@
       "version": "~2.0.2-SNAPSHOT"
     },
     "org.zowe.explorer-ip": {
-      "version": "~1.0.0-SNAPSHOT",
+      "version": "~1.0.1-SNAPSHOT",
```
Comment (1000TurquoisePogs) [inline manifest.json.template]: 2.0.0

**Review by struga0258 — APPROVED** [review]

**Pull request merged**